core/src/main/java/org/apache/iceberg/BaseMetastoreTableOperations.java

```diff
@@ -174,7 +174,7 @@ protected void refreshFromMetadataLocation(String newLocation, int numRetries) {
   protected void refreshFromMetadataLocation(String newLocation, Predicate<Exception> shouldRetry,
                                              int numRetries) {
     refreshFromMetadataLocation(newLocation, shouldRetry, numRetries,
-        metadataLocation -> TableMetadataParser.read(io(), metadataLocation));
+        metadataLocation -> TableMetadata.buildFromLocation(io(), metadataLocation).build());
```

Comment:
I don't think we should change the metadata after reading it, so it is probably best to fix this up in the `TableMetadata` constructor rather than in the builder, which is used to produce new `TableMetadata` objects.